### PR TITLE
function to read compacted topics

### LINKED ISSUE #5538: Pulsar function to read compacted topics
**Is your feature request related to a problem? Please describe.**
Pulsar functions cannot currently be configured to read the compacted version of a topic.

**Describe the solution you'd like**
An option to `pulsar-admin functions create/update`, and/or setting in `--function-config-file`, to request subscribing to the compacted version of a topic.

Note: a function can have multiple topics as inputs.  There are already other options which are on a per-input basis (e.g. `custom-schema-inputs`, `custom-serde-inputs`) so it ought to work like those.

**Describe alternatives you've considered**
It occurred to me that subscription options could be part of the topic URI, e.g.

```
persistent://public/default/my-topic?compacted=yes
```

... but that would be a major change if deployed everywhere that subscriptions are used.

**Additional context**
A similar option might be required for sinks.

PR #5532 adds a similar feature to start a sink from "earliest" rather than "latest" message.  If that's just for sinks then it would be useful for functions too.

Hence there seems to be a more general case use around setting initial subscription options for both functions and sinks.

### PR DESCRIPTION
**Pull request by 315157973**
Fixes #5538

### Modifications
In function mode and sink mode, PulsarSource can read compacted topic。
By `inputSpecs` parameter, each topic can independently decide whether to read compacted

### Verifying this change
unit test:
org.apache.pulsar.io.PulsarFunctionE2ETest#testReadCompactedFunction
org.apache.pulsar.io.PulsarFunctionE2ETest#testReadCompactedSink

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java [hunk 1]
```diff
@@ -233,6 +234,8 @@ void processArguments() throws Exception {
         protected String customSerdeInputString;
         @Parameter(names = "--custom-schema-inputs", description = "The map of input topics to Schema class names (as a JSON string)")
         protected String customSchemaInputString;
+        @Parameter(names = "--input-specs", description = "The map of inputs to custom configuration (as a JSON string)")
```
Comment (codelipenghui) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java:237]: Is this related to this PR?

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdSinks.java [hunk 2]
```diff
@@ -276,6 +277,9 @@ protected void validateSinkConfigs(SinkConfig sinkConfig) {
         @Parameter(names = "--custom-schema-inputs", description = "The map of input topics to Schema types or class names (as a JSON string)")
         protected String customSchemaInputString;
 
+        @Parameter(names = "--input-specs", description = "The map of inputs to custom configuration (as a JSON string)")
```
Comment (codelipenghui) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdSinks.java:280]: Is this related to this PR?

File: pulsar-functions/proto/src/main/proto/Function.proto [hunk 3]
```diff
@@ -94,6 +94,7 @@ message ConsumerSpec {
         int32 value = 1;
     }
     ReceiverQueueSize receiverQueueSize = 4;
+    bool readCompacted = 5;
```
Comment (sijie) [inline pulsar-functions/proto/src/main/proto/Function.proto:97]: I think a more generic approach is to add a `Map<String, String>` to the consumer spec. So we can use https://github.com/apache/pulsar/blob/870a637b4906862a611e418341dd926e21458f08/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBuilderImpl.java#L86 to load the consumer configuration when building the consumers.

Comment (jerrypeng) [reply pulsar-functions/proto/src/main/proto/Function.proto:97]: We already have a ConsumerConfig class:

https://github.com/apache/pulsar/blob/master/pulsar-common/src/main/java/org/apache/pulsar/common/functions/ConsumerConfig.java

Though I am for refactoring this to be generic Map so we don't have to make code changes for new consumer configs that are added.

Comment (315157973) [reply pulsar-functions/proto/src/main/proto/Function.proto:97]: > We already have a ConsumerConfig class:
> 
> https://github.com/apache/pulsar/blob/master/pulsar-common/src/main/java/org/apache/pulsar/common/functions/ConsumerConfig.java
> 
> Though I am for refactoring this to be generic Map so we don't have to make code changes for new consumer configs that are added.

Thanks very much. But schemaProperties are properties belonging to the schema. Put the Consumer parameters with them, I don’t think it’s a good way. I think it is necessary to add a new Map to save the general parameters of Consumer.And this make it easier to convert Map to object

**Comment by jiazhai** [comment]
@jerrypeng @sijie Would you please help review this PR again?

**Comment by jdimeo** [comment]
How does one configure their function to read from a compacted topic? the current documentation doesn't seem to mention it and it's not obvious looking at the code of this PR. Does it happen automatically? does `readCompacted: true` go into the user config properties?
